### PR TITLE
Making VTK reader more versatile

### LINKED ISSUE #2001: VTK reader
There are a couple errors with the VTK reader. These are all with the ASCII VTK format.

1. The ASCII format should allow arbitrary blank lines but if there is a blank line after `ASCII` (i.e. if line 4 is blank), the reader crashes and an error is thrown.
```
MFEM abort: VTK mesh is not UNSTRUCTURED_GRID!
 ... in function: void mfem::Mesh::ReadVTKMesh(std::istream&, int&, int&, bool&)
 ... in file: /home/wcdawn/work/microreactor/mfem/mesh/mesh_readers.cpp:400
```

2. The `CELL_DATA` array `material` **must** be the first array on the file. If any `POINT_DATA` or other `CELL_DATA` arrays are specified, the material is treated as homogeneous. The VTK reader should simply skip over any other arrays besides the `material` array.

I realize this may not be a much used feature but it would be nice to have these bugs fixed. I can provide an example VTK if that would help.

### PR DESCRIPTION
**Pull request by wcdawn**
This allows blank lines in the header block of the VTK file and allows the `material` data to appear anywhere in the file allowed by the VTK format.

Resolves #2001 

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2028](https://github.com/mfem/mfem/pull/2028) | @wcdawn | @tzanio | @pazner + @bslazarov + @mlstowell | 02/07/21 | 02/22/21 | 03/02/21 | |

### CODE REVIEW

**Comment by pazner** [comment]
```diff
diff --git a/mesh/mesh_readers.cpp b/mesh/mesh_readers.cpp
index b67105aee..0d7a664e8 100644
--- a/mesh/mesh_readers.cpp
+++ b/mesh/mesh_readers.cpp
@@ -786,17 +786,13 @@ void Mesh::ReadVTKMesh(std::istream &input, int &curved, int &read_gf,
       MFEM_ABORT("VTK mesh is not in ASCII format!");
       return;
    }
-   buff = "";
-   while (buff == "") // allow blanklines
+   do
    {
       getline(input, buff);
       filter_dos(buff);
+      if (!input.good()) { MFEM_ABORT("VTK mesh is not UNSTRUCTURED_GRID!"); }
    }
-   if (buff != "DATASET UNSTRUCTURED_GRID")
-   {
-      MFEM_ABORT("VTK mesh is not UNSTRUCTURED_GRID!");
-      return;
-   }
+   while (buff != "DATASET UNSTRUCTURED_GRID");
 
    // Read the points, skipping optional sections such as the FIELD data from
    // VisIt's VTK export (or from Mesh::PrintVTK with field_data==1).

```

File: mesh/mesh_readers.cpp [hunk 1]
```diff
@@ -862,30 +866,39 @@ void Mesh::ReadVTKMesh(std::istream &input, int &curved, int &read_gf,
       cell_types.Load(ncells, input);
```
Comment (pazner) [inline mesh/mesh_readers.cpp:866]: Not strictly related to this PR, but the `CELL_TYPES` are actually mandatory, so we should error if we don't find them:
```diff
diff --git a/mesh/mesh_readers.cpp b/mesh/mesh_readers.cpp
index b67105aee..cc9ec80e4 100644
--- a/mesh/mesh_readers.cpp
+++ b/mesh/mesh_readers.cpp
@@ -860,11 +860,9 @@ void Mesh::ReadVTKMesh(std::istream &input, int &curved, int &read_gf,
    input >> ws >> buff;
    Array<int> cell_types;
    int ncells;
-   if (buff == "CELL_TYPES")
-   {
-      input >> ncells;
-      cell_types.Load(ncells, input);
-   }
+   MFEM_VERIFY(buff == "CELL_TYPES", "CELL_TYPES not provided in VTK mesh.")
+   input >> ncells;
+   cell_types.Load(ncells, input);
 
    while ((input.good()) && (buff != "CELL_DATA"))
    {
```

File: mesh/mesh_readers.cpp [hunk 2]
```diff
@@ -862,30 +866,39 @@ void Mesh::ReadVTKMesh(std::istream &input, int &curved, int &read_gf,
...
       }
    }
-   else
+
+   if (!found_material)
    {
-      input.seekg(sp);
+      MFEM_WARNING("Material array not found in VTK file. "
+                   "Assuming uniform material composition.");
    }
```
Comment (pazner) [inline mesh/mesh_readers.cpp:902]: I don't think we need to warn if the materials/attributes are missing.

**Comment by pazner** [comment]
Looks good modulo a few minor comments. It will be useful to have a more robust VTK reader, thanks @wcdawn.

File: mesh/mesh_readers.cpp (hunk 2 above)
Comment (wcdawn) [reply mesh/mesh_readers.cpp:896]: It was definitely useful for debugging this but I'm alright with removing it. Either way.

**Comment by wcdawn** [comment]
Thanks for the comments! They helped clarify the code so I've applied the diffs.

**Review by bslazarov — APPROVED** [review]

**Comment by bslazarov** [comment]
Do we have a test input vtk file? It will be nice to add one if we do not have it.

**Comment by wcdawn** [comment]
I've attached an example VTK file that I use. You'll need to change the extension from `txt` to `vtk`.
My typical problems have many materials and not as much interesting geometry.
I can add a unit test if that is wanted.
[iaea_refl_sixth.txt](https://github.com/mfem/mfem/files/6011226/iaea_refl_sixth.txt)

File: mesh/mesh_readers.cpp (hunk 2 above)
Comment (pazner) [reply mesh/mesh_readers.cpp:896]: Ok, my vote is for removing the warning… @bslazarov, @mlstowell, @tzanio?

**Review by pazner — APPROVED** [review]

File: mesh/mesh_readers.cpp (hunk 2 above)
Comment (tzanio) [reply mesh/mesh_readers.cpp:896]: I think we can leave it commented out

**Comment by tzanio** [comment]
@mlstowell, do you want to take a look at this?

**Review by mlstowell — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Review by tzanio — APPROVED** [review]

**Comment by v-dobrev** [comment]
@tzanio, the last commit from this branch is not in `next` and the Travis CI builds of `next` are still failing.

**Comment by tzanio** [comment]
I know, I'll fix it

**Pull request merged**